### DISCUSSION TITLE
MIDI button on mouse down only?  (Panic button)

### DISCUSSION THREAD

**Discussion by Andy2No** [RomanKubiak/ctrlr#664] [Q&A] [answered]
How do I make a button that sends a CC on mouse down, but not on mouse up?

I'm trying to send CC 123, (MIDI Panic / all notes off).  AFAIK, the data byte doesn't matter.
So far, I have a uiButton, with "Trigger button on mouse down events" ticked, that's sending a CC 123 when I click it, then another when I left go.  I can change the data value that's sent, but what I really want to do is _only_ send CC 123 when I click on the button, and not when I let go.

Is that possible?

I'm using Ctrlr Version = 5.3.201, Build date = Wed Mar 30 00:52:44 CEDT 2016, Branch = Nightly, Juce = 4.0.2, libusb = 1.0.19, liblo = 0.28, 

Also, the button doesn't always work - I often have to click on one of the other controls first (all the others are type uiSlider).  I assume this is a bug in the version I'm using, so perhaps there's a better version?

EDIT:  I've attached the zipped panel - there are no images in it.  It's just a couple of knobs for fine pitch bend and ten knobs to send CCs, then the Panic button below that.

[Scratch Pad_0_6_Experimental_2024-07-09_02-58.zip](https://github.com/user-attachments/files/16136135/Scratch.Pad_0_6_Experimental_2024-07-09_02-58.zip)

**Comment by dnaldoog** [comment, marked as answer]
If you uncheck that option "__Trigger button on mouse down events__" and set the button as a toggle  - then the button sends out one message, but it seems to be on mouseUp, but that should be okay, right?
####

![image](https://github.com/RomanKubiak/ctrlr/assets/76831683/09402d07-eaaf-42c1-8929-211c53094f16)


If you specifically need a mouse down event ,then you would need to create a lua callback in the mouseDown field in the editor remembering to disable the CC in the editor:
####

![image](https://github.com/RomanKubiak/ctrlr/assets/76831683/352012bb-75d4-4a52-bb68-cdb9bc5c0246)


![image](https://github.com/RomanKubiak/ctrlr/assets/76831683/dd9e19ae-0eae-428d-acb2-74deb255d12b)

<pre>
<code>
mouseDown = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)
    local channelOut = panel:getProperty("panelMidiOutputChannelDevice")
    local statusByte = 0xB0 + (channelOut - 1) -- use this code for dynamic channel changes
	local message=MemoryBlock()
	message:loadFromHexString(string.format("%.2X %.2x %.2X",statusByte,123,64))
	panel:sendMidiMessageNow(CtrlrMidiMessage(message:toHexString(1)))
end
</code>
</pre>

####
[__EDIT__ :Scratch Pad_0_6_Experimental_mousedown2.zip](https://github.com/user-attachments/files/16152186/Scratch.Pad_0_6_Experimental_mousedown2.zip)

**Reply by Andy2No** [reply]
Excellent.  Thanks for the explanation and for going the extra mile and making the button for me, @dnaldoog.  This one seems 100% reliable so far too - I haven't had to click on one of the sliders first to get it to respond.

I managed to make a Lua method for something, just once, but I'm a long way from being comfortable doing it.  It seems like a useful thing to learn though.  I get lost in all the form based settings, not fully understanding what a lot of them are for, and often struggling to find the same one again, even the same day.  Doing more of it as code might be easier, once I know how.

**Comment by dnaldoog** [comment]
Great. Yes lua is well worth getting into if you decide to make your panel do more, like save and load patches or process sysex dumps etc.

I have the same experience- constantly having to go and learn something I have forgotten. I think/hope it’s pretty normal!
